core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesUnassignedShardsAction.java

```diff
@@ -0,0 +1,179 @@
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeReadAction;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.MutableShardRouting;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.gateway.AsyncShardFetch;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Transport action that reads the cluster state for any unassigned shards for specific indices
+ * and fetches the shard version, host node id and any exception while opening the shard index
+ * from all the nodes.
+ */
+public class TransportIndicesUnassignedShardsAction extends TransportMasterNodeReadAction<IndicesUnassignedShardsRequest, IndicesUnassigedShardsResponse> {
+
+    private final TransportNodesListGatewayStartedShards listShardsInfo;
+
+    @Inject
+    public TransportIndicesUnassignedShardsAction(Settings settings, TransportService transportService, ClusterService clusterService, ThreadPool threadPool, ActionFilters actionFilters,
+                                                  TransportNodesListGatewayStartedShards listShardsInfo) {
+        super(settings, IndicesUnassignedShardsAction.NAME, transportService, clusterService, threadPool, actionFilters, IndicesUnassignedShardsRequest.class);
+        this.listShardsInfo = listShardsInfo;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected IndicesUnassigedShardsResponse newResponse() {
+        return new IndicesUnassigedShardsResponse();
+    }
+
+    @Override
+    protected void masterOperation(IndicesUnassignedShardsRequest request, ClusterState state, ActionListener<IndicesUnassigedShardsResponse> listener) {
+        // collect all unassigned shards for the requested indices
+        DiscoveryNodes nodes = state.nodes();
+        MetaData metaData = state.metaData();
+        Set<String> requestedIndices = new HashSet<>();
+        requestedIndices.addAll(Arrays.asList(request.indices()));
+        List<ShardId> shardIdsToFetch = new ArrayList<>(state.routingNodes().unassigned().size());
+        for (MutableShardRouting shard : state.routingNodes().unassigned()) {
```

Comment:
since you are executing this on the master I think you should also take `state.routingNodex().ignoredUnassigned()` into account. Those shards are also interesting. I wonder if we should _flag_ them somehow special but we can do that later...